external-crates/move/crates/move-compiler/src/sui_mode/info.rs

```diff
@@ -0,0 +1,304 @@
+                    .filter(|(_t, is_phantom)| *is_phantom)
+                    .map(|(t, _is_phantom)| visit_ty(info, visited, uid_holders, t))
+                    .fold(None, merge_uid_holder_opt);
+                let tn_holder = if let N::TypeName_::ModuleType(m, n) = tn.value {
+                    visit_decl(info, visited, uid_holders, m, n);
+                    uid_holders.get(&(m, n)).copied()
+                } else {
+                    None
+                };
+                merge_uid_holder_opt(ty_args_holder, tn_holder)
+            }
+        }
+    }
+
+    fn visit_fields(
+        info: &TypingProgramInfo,
+        visited: &mut BTreeSet<(ModuleIdent, DatatypeName)>,
+        uid_holders: &mut BTreeMap<(ModuleIdent, DatatypeName), UIDHolder>,
+        fields: &Fields<N::Type>,
+    ) -> Option<UIDHolder> {
+        fields
+            .key_cloned_iter()
+            .map(|(field, (_, ty))| {
+                Some(match visit_ty(info, visited, uid_holders, ty)? {
+                    UIDHolder::IsUID => UIDHolder::Direct { field, ty: ty.loc },
+                    UIDHolder::Direct { field, ty: uid }
+                    | UIDHolder::Indirect { field, uid, ty: _ } => UIDHolder::Indirect {
+                        field,
+                        ty: ty.loc,
+                        uid,
+                    },
+                })
+            })
+            .fold(None, merge_uid_holder_opt)
+    }
+
+    fn visit_decl(
+        info: &TypingProgramInfo,
+        visited: &mut BTreeSet<(ModuleIdent, DatatypeName)>,
+        uid_holders: &mut BTreeMap<(ModuleIdent, DatatypeName), UIDHolder>,
+        mident: ModuleIdent,
+        tn: DatatypeName,
+    ) {
+        if visited.contains(&(mident, tn)) {
+            return;
+        }
+        visited.insert((mident, tn));
+
+        let uid_holder_opt = match info.datatype_kind(&mident, &tn) {
+            DatatypeKind::Struct => match &info.struct_definition(&mident, &tn).fields {
+                N::StructFields::Defined(_, fields) => {
+                    visit_fields(info, visited, uid_holders, fields)
+                }
+                N::StructFields::Native(_) => None,
+            },
+            DatatypeKind::Enum => info
+                .enum_definition(&mident, &tn)
+                .variants
+                .iter()
+                .filter_map(|(_, _, v)| match &v.fields {
+                    N::VariantFields::Defined(_, fields) => Some(fields),
+                    N::VariantFields::Empty => None,
+                })
+                .map(|fields| visit_fields(info, visited, uid_holders, fields))
+                .fold(None, merge_uid_holder_opt),
+        };
+        if let Some(uid_holder) = uid_holder_opt {
+            uid_holders.insert((mident, tn), uid_holder);
+        }
+    }
+
+    // iterate over all struct/enum declarations
+    let visited = &mut BTreeSet::new();
+    let mut uid_holders = BTreeMap::new();
```

Comment:
Nit: why not `let mut visited = BTreeSet::new()`?

Reply:
Where?

Reply:
instead of `let visited = &mut BTreeSet::new()`, doing `let mut visited = BTreeSet::new()`. But it's more of a stylistic nit so feel free to ignore if you prefer it this way.

Reply:
I like avoiding the owned value if I can :) Less typing